sources/network/unix/io_service.cpp

```diff
@@ -57,23 +63,56 @@ set_default_io_service(const std::shared_ptr<io_service>& service) {
 //! ctor & dtor
 //!
 
-io_service::io_service(void)
-: m_should_stop(false)
-, m_callback_workers(__TACOPIE_IO_SERVICE_NB_WORKERS) {
+io_service::io_service(std::size_t nb_threads, bool bDelayedStart)
+: m_should_stop(ATOMIC_VAR_INIT(false))
+, m_notifier(bDelayedStart)
+, m_callback_workers(nb_threads) {
   __TACOPIE_LOG(debug, "create io_service");
 
-  //! Start worker after everything has been initialized
-  m_poll_worker = std::thread(std::bind(&io_service::poll, this));
+  if (!bDelayedStart)
+    startup();
 }
 
 io_service::~io_service(void) {
   __TACOPIE_LOG(debug, "destroy io_service");
 
+  shutdown();
+}
+
+void
+io_service::startup(void) {
+  __TACOPIE_LOG(debug, "starting io_service");
+
+#ifdef _WIN32
+  //! Start winsock before any other socket calls.
+  WSADATA wsaData;
+  int nRet = WSAStartup(0x202, &wsaData);
+  if (nRet)
+    throw cpp_redis::redis_error("Could not startup io_service, WSAStartup() failure");
+#endif
+
+  m_notifier.startup();
+
+  //! Start worker after everything has been initialized
+  m_poll_worker = std::thread(std::bind(&io_service::poll, this));
+}
+
+void
+io_service::shutdown(void) {
+  __TACOPIE_LOG(debug, "shutting down io_service");
+
   m_should_stop = true;
 
   m_notifier.notify();
-  m_poll_worker.join();
+  if (m_poll_worker.joinable())
+    m_poll_worker.join();
   m_callback_workers.stop();
+
+  m_notifier.shutdown();
+
+#ifdef _WIN32
+  WSACleanup();
```

Comment:
should not be handle by tacopie itself, though more convenient. This might break app if multiple io_service or multiple networking libraries.